test/replication-luatest/gh_6842_qsync_applier_order_test.lua

```diff
@@ -162,3 +186,91 @@ g.test_local_txn_during_remote_promote = function(g)
     end)
     luatest.assert_equals(content, {{3}}, 'synchro transactions work')
 end
+
+--
+-- Server 1 was a synchro queue owner. It starts a synchro txn. The txn is
+-- written to WAL but not reported to TX yet. Relay manages to send it. Server 2
+-- receives the txn and makes a PROMOTE including this txn.
+--
+-- Then the PROMOTE is delivered to server 1 and goes to WAL too. Then the txn
+-- and PROMOTE WAL writes are processed by TX thread in a single batch without
+-- yields.
+--
+-- The bug was that the txn's synchro queue entry didn't get an LSN right after
+-- WAL write and PROMOTE couldn't process synchro entries not having an LSN.
+--
+g.test_remote_promote_during_local_txn_including_it = function(g)
+    -- Start synchro txns on server 1.
+    local fids = g.server1:exec(function()
+        box.ctl.promote()
+        local s = box.schema.create_space('test', {is_sync = true})
+        s:create_index('pk')
+        box.cfg{
+            -- To hang own transactions in the synchro queue.
+            replication_synchro_quorum = 3,
+        }
+        box.error.injection.set('ERRINJ_RELAY_FASTER_THAN_TX', true)
+        local fiber = require('fiber')
+        -- More than one transaction to ensure that it works not just for one.
+        local f1 = fiber.new(s.replace, s, {1})
+        f1:set_joinable(true)
+        local f2 = fiber.new(s.replace, s, {2})
+        f2:set_joinable(true)
+        return {f1:id(), f2:id()}
+    end)
+
+    -- The txns are delivered to server 2.
+    wait_synchro_len(g.server2, #fids)
+
+    -- Server 2 confirms the txns and sends a PROMOTE covering it to server 1.
+    g.server2:exec(function()
+        box.cfg{
+            replication_synchro_quorum = 1,
+            -- To make it not wait for confirm from server 1. Just confirm via
+            -- the promotion ASAP.
+            replication_synchro_timeout = 0.001,
+        }
+        box.ctl.promote()
+        box.cfg{
+            replication_synchro_quorum = 2,
+            replication_synchro_timeout = 1000,
+        }
+    end)
+
+    -- Server 1 receives the foreign PROMOTE. Now the local txns and a remote
+    -- PROMOTE are being written to WAL.
+    wait_synchro_is_busy(g.server1)
+
+    local rows = g.server1:exec(function(fids)
+        -- Simulate the TX thread being slow. To increase likelihood of the txn
+        -- and PROMOTE WAL writes being processed by TX in a single batch.
+        box.error.injection.set('ERRINJ_TX_DELAY_PRIO_ENDPOINT', 0.1)
```

Comment:
Ok, going back to that ERRINJ.
Seems if you used ERROR_INJECT_SLEEP, you would still have to wait for some timeout here in the test to make sure everyone is stuck in the same tx_prio process loop.
Correct?

Reply:
I suppose you are talking about a 100% reproducer when I would explicitly wait for stuff being blocked? I intentionally didn't bother with that. The idea is to just use a big enough timeout to guarantee that on any machine a couple of WAL writes could be finished. To get a 'good enough' reproducer. Otherwise the problem is that to guarantee stuff is blocked you would need to block TX thread + wait for WAL thread to do stuff. But while TX is blocked, you can't wait, because the polling happens also in TX thread. A possible solution would be to stack finished WAL entries in WAL thread in a separate queue and prevent its flush while another errinj is set + via a third errinj we track its size. However sounds too complicated.

Reply:
Yep, I missed that the TX thread couldn't be polled while that errinj is set. Thanks for the explanation!